### PR TITLE
Uses direct routing when direct channel exists, no pfs request

### LINKED ISSUE #3303: Only ask PFS if necessary
## Abstract

With #3236 Raiden will ask the PFS for a path if the PFS is set. However, this is not necessary for transfers to direct channel partners. In this case the Raiden client shouldn't ask the PFS:

## Motivation

Not doing the request to the PFS will lower costs and increase latency of payments to direct partners.

## Specification

Before asking the PFS the client should check if:
- there is a direct channel to the target
- if this channel is open
- if the channel has enough capacity

If all checks are successful the payment can be sent using this channel.

## Backwards Compatibility

Backwards compatible.

### PR DESCRIPTION
**Pull request by Dominik1999**
closes #3303

### CODE REVIEW

File: raiden/routing.py [hunk 1]
```diff
@@ -36,7 +36,20 @@ def get_best_routes(
...
+            chain_state=chain_state,
+            token_network_id=token_network_id,
+            from_address=from_address,
+            to_address=to_address,
+            amount=amount,
+            previous_address=previous_address,
+        )
+        for route_state in neighbours:
+            if to_address == route_state.node_address:
+                return [route_state]
```
Comment (palango) [inline raiden/routing.py]: @rakanalh @hackaugusto When doing a transfer we want to use the direct channel if possible. However, the capacity of the channel will be checked later, so this will not work when the channel doesn't have sufficient capacity. Is it OK to check the capacity before?

File: raiden/tests/utils/factories.py [hunk 2]
```diff
@@ -982,8 +982,6 @@ def create_network(
             block_hash=block_hash,
         )
         state = iteration.new_state
-
```
Comment (palango) [inline raiden/tests/utils/factories.py]: Please undo these changes.

File: raiden/routing.py (hunk 1 above)
Comment (rakanalh) [reply raiden/routing.py]: The reason we did not check capacity in the first place was the fact that we wanted to have (at some point) the transfer wait until both enough capacity / node availability conditions are met for the mediators to forward the transfer, as capacity/availability might change during the transfer's lifecycle. We've implemented certain state changes to notify about availability but not capacity when we removed the checks here in the routing module but not capacity change notifications. For now, i guess there's no harm in restoring checks for capacity until we decide to remove it.

Comment (Dominik1999) [reply raiden/routing.py]: Should we introduce the capacity check for all the nodes again or just for the direct transfer?

File: raiden/routing.py [hunk 3]
```diff
@@ -36,6 +36,32 @@ def get_best_routes(
...
+            previous_address=previous_address,
+        )
+        channel_state = views.get_channelstate_by_token_network_and_partner(
+            chain_state=chain_state,
+            token_network_id=token_network_id,
+            partner_address=Address(to_address),
+        )
+
+        for route_state in neighbours:
+            if to_address == route_state.node_address:
```
Comment (palango) [inline raiden/routing.py]: You can combine these two ifs

File: raiden/routing.py [hunk 4]
```diff
@@ -36,6 +36,32 @@ def get_best_routes(
...
+                if (
+                    channel_state
+                    and channel.get_distributable(
```
Comment (palango) [inline raiden/routing.py]: Please add a comment describing that the other conditions (channel is open, ...) are already checked by `get_best_routes_internal`

File: raiden/routing.py [hunk 5]
```diff
@@ -36,6 +36,32 @@ def get_best_routes(
 ) -> List[RouteState]:
     services_config = config.get("services", None)
 
+    if to_address in views.all_neighbour_nodes(chain_state):
```
Comment (palango) [inline raiden/routing.py]: Please add a comment here describing what and why we do this.

File: raiden/tests/unit/test_pfs_integration.py [hunk 6]
```diff
@@ -273,9 +280,12 @@ def test_routing_mocked_pfs_request_error(chain_state, token_network_state, our_
             chain_state=chain_state,
             token_network_state=token_network_state,
             from_address=our_address,
-            to_address=address1,
+            to_address=address4,
             amount=50,
         )
+        # PFS doesn't work, so internal routing is used, so two possible routes are returned,
```
Comment (palango) [inline raiden/tests/unit/test_pfs_integration.py]: 👍

File: raiden/tests/unit/test_pfs_integration.py [hunk 7]
```diff
@@ -296,11 +306,18 @@ def test_routing_mocked_pfs_bad_http_code(chain_state, token_network_state, our_
         address3: NODE_NETWORK_REACHABLE,
     }
 
-    # channel 1 and 2 are flipped here, to see when the PFS gets called
+    # in case the pfs sends a bad hhtp code but the correct path back
```
Comment (palango) [inline raiden/tests/unit/test_pfs_integration.py]: http code

Comment (palango) [inline raiden/tests/unit/test_pfs_integration.py]: Is getting the correct path but a wrong code something that is tested here?

Comment (Dominik1999) [reply raiden/tests/unit/test_pfs_integration.py]: in that test, we mock the response code to 400 and the answer as the correct path. We see in the asserts that internal routing is used due to the bad code, right?

Comment (palango) [reply raiden/tests/unit/test_pfs_integration.py]: Makes sense!

Comment (Dominik1999) [reply raiden/tests/unit/test_pfs_integration.py]: Should we remove that test?

Comment (palango) [reply raiden/tests/unit/test_pfs_integration.py]: No, it makes sense.

**Comment by palango** [comment]
@rakanalh PR is fine from my side. I wonder if we should merge it before the next mainnet release.

On the one hand I think the changes are clear and well tested, but on the other hand it changes behavior a bit. I tend to merge it.

**Comment by rakanalh** [comment]
Skimming through the PR, adding the capacity check is also fine by me.

I'll look into the tests and review the PR today

**Comment by Dominik1999** [comment]
Guys, the PR is already reviewed from Paul I guess. Just want to prevent that you do double the work

**Comment by palango** [comment]
Yes, the question was more aimed at whether or not we're fine to change the behaviour of client for the mainnet release.

**Comment by rakanalh** [comment]
I don't see anything that could prevent this from being included in the mainnet release. So if Paul is fine with the PR, let's merge it.
